sdk/cosmos/azure-cosmos/azure/cosmos/_container_recreate_retry_policy.py

```diff
@@ -0,0 +1,165 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""Internal class for container recreate retry policy implementation in the Azure
+Cosmos database service.
+"""
+import ast
+import json
+
+from typing import Optional, Dict, Any, List, Union
+from . import http_constants
+from .partition_key import _Empty, _Undefined
+
+
+# pylint: disable=protected-access
+
+
+class ContainerRecreateRetryPolicy:
+    def __init__(self, client: Optional[Any], container_caches: Optional[Dict[str, Dict[str, Any]]],
+                 headers: Dict[str, Any], *args: Optional[List]):
+        self.retry_after_in_milliseconds = 0  # Same as in .net
+        self.refresh_container_properties_cache = True
+        self.args = args
+        self._intended_headers = http_constants.HttpHeaders.IntendedCollectionRID
+        self.container_rid = None
+        self.container_link = None
+        self.link = None
+        self._headers = headers
+        if self._headers and self._intended_headers in self._headers:
+            self.container_rid = self._headers[self._intended_headers]
+            if container_caches:
+                self.container_link = self.__find_container_link_with_rid(container_caches, self.container_rid)
+        self.client = client
+        self.exception = None
+
+    def ShouldRetry(self, exception: Optional[Any]) -> bool:
+        """Returns true if the request should retry based on the passed-in exception.
+
+        :param (exceptions.CosmosHttpResponseError instance) exception:
+        :returns: a boolean stating whether the request should be retried
+        :rtype: bool
+
+        """
+
+        self.exception = exception  # needed for pylint
+        if self.refresh_container_properties_cache:
+            if not self.container_rid or not self.container_link:
+                return False
+            self.refresh_container_properties_cache = False
+            return True
+        return False
+
+    def __find_container_link_with_rid(self, container_properties_caches: Optional[Dict[str, Any]], rid: str) -> \
```

Comment:
nit: The caller uses `Optional[Dict[str, Dict[str, any]]]` for `container_caches` so the nested type is `Dict` - I wonder why do we need to iterate and do a value match of `rid` - can't we do a direct access of the appropriate key to fetch `rid`?

Reply:
That is because the container link is the key, we don't pass it when we get to this point and only have access to the RID that resulted in a mismatch (container recreate) so we use the RID to get the container link so we can easily access the container properties needed to do a retry. We only have to do this once during the retry.